src/distrs/beta.jl

```diff
@@ -1,19 +1,71 @@
 # functions related to beta distributions
 
+using HypergeometricFunctions: _₂F₁
+
 import .RFunctions:
-    betapdf,
-    betalogpdf,
-    betacdf,
-    betaccdf,
-    betalogcdf,
-    betalogccdf,
-    betainvcdf,
-    betainvccdf,
+    # betapdf,
+    # betalogpdf,
+    # betacdf,
+    # betaccdf,
+    # betalogcdf,
+    # betalogccdf,
+    # betainvcdf,
+    # betainvccdf,
     betainvlogcdf,
     betainvlogccdf
 
 # pdf for numbers with generic types
-betapdf(α::Real, β::Real, x::Number) = x^(α - 1) * (1 - x)^(β - 1) / beta(α, β)
+betapdf(α::Real, β::Real, x::Real) = exp(betalogpdf(α, β, x))
 
 # logpdf for numbers with generic types
-betalogpdf(α::Real, β::Real, x::Number) = (α - 1) * log(x) + (β - 1) * log1p(-x) - logbeta(α, β)
+function betalogpdf(α::T, β::T, x::T) where T<:Real
+    return if x <= 0.5
+        xlogy(α - 1, x) + (β - 1) * log1p(-x) - logbeta(α, β)
+    else
+        betalogpdf(β, α, 1 - x)
+    end
+end
+betalogpdf(α::Real, β::Real, x::Real) = betalogpdf(promote(α, β, x)...)
+
+betacdf(α::Float64, β::Float64, x::Float64) = first(beta_inc(α, β, x))
+betacdf(α::Real, β::Real, x::Real) = betacdf(promote(float(α), β, x)...)
+betacdf(α::T, β::T, x::T) where T = throw(MethodError(betacdf, (α, β, x)))
+
+betaccdf(α::Float64, β::Float64, x::Float64) = last(beta_inc(α, β, x))
+betaccdf(α::Real, β::Real, x::Real) = betaccdf(promote(float(α), β, x)...)
+betaccdf(α::T, β::T, x::T) where T = throw(MethodError(betaccdf, (α, β, x)))
```

Comment:
I think we'd still need a promoting version somewhere. Either here or in SpecialFunctions.

Reply:
Sure, it's needed somewhere. To me it just seems it's better to have it in SpecialFunctions since it is not only of interest to StatsFuns. For the 4-argument version of `beta_inc`, these promotions (and some additional ones) are already contained in SpecialFunctions. I opened https://github.com/JuliaMath/SpecialFunctions.jl/pull/318 to extend them to the 3-argument version that is called here.

Reply:
I missed that method when looking at your PR. Yes, let's just have the promotion there and simplify things here.